### ISSUE TITLE
Do not allow varchar of length zero

### ISSUE THREAD

**Issue by electrum** [prestosql/presto#1136]
From the SQL standard:
```
6.1 <data type>
Syntax Rules
4) The value of a <length> shall be greater than 0 (zero).
```

PostgreSQL fails:
```
# create table foo (x varchar(0));
ERROR:  length for type varchar must be at least 1
LINE 1: create table foo (x varchar(0));
```

We should fix computations of an empty string to be a legal `varchar(1)` and disallow the creation of a varchar with a zero length.

**Comment by hustnn** [comment]
@electrum 
Presto already disallowed the creation of a varchar with a zero length.
`create table foo_test_varchar (x varchar(0));
Query 20190717_032945_07332_rx6e3 failed: Varchar length 0 out of allowed range [1, 65535]`

Seems only need to a PR to fix the computations of an empty string to be `varchar(1)` or simply a `varchar`, right?

**Comment by electrum** [comment]
That looks like a connector specific error message. The Presto engine allows varchar up to `MAX_INT`. The validation in `VarcharType` needs to be changed to disallow zero, along with empty string and any other places that would generate it (or be left alone to fail naturally if it’s the result of bad user input).

**Comment by sopel39** [comment]
Does it also apply to `Char` type?

**Comment by hustnn** [comment]
@electrum 
Changing the validation in `VarcharType` is not easy, but seems a lot of intermediate expression evaluation in class like `ExpressionAnalyzer` will be impacted. The intermediate expression `''` with type `VARCHAR(0)` should be legal in expression `REPLACE('foo', '')`, right?

**Comment by electrum** [comment]
@sopel39 yes, that section covers `char`, `varchar`, `binary`, `varbinary`, and others

@hustnn That's a good question. What type should `'x' || '' || 'y'` produce? Today, it produces `varchar(2)`, but changing `''` to `varchar(1)` would cause it to produce `varchar(3)`. Maybe that's ok. I'm not sure if our current type calculation is mandated by the specification or if that part is implementation defined. @martint do you remember?

**Comment by findepi** [comment]
@electrum @sopel39 i am under impression that string literals (`'...'`) should have `char(x)` type. Do they have `varchar(x)` type for some legacy reasons?
If that's the case, type of `''` should be `char(0)`.

**Comment by sopel39** [comment]
For char there might be data sources that don't comply with the standard. In such case we probably shouldn't map input zero-length chars to Presto `char(1)` as we change data.

**Comment by findepi** [comment]
`''` is definitely _not_ `char(1)`. If it's not `char(0)`, then it must be `varchar`.

**Comment by martint** [comment]
The SQL spec is not very clear about zero-length strings:

```
17) The declared type of a <character string literal> is fixed-length character string. The length of a <character string literal> is the number of <character representation>s that it contains. Each <quote symbol> contained in <character string literal> represents a single <quote> in both the value and the length of the <character string literal>. The two <quote>s contained in a <quote symbol> shall not be separated by any <separator>.

NOTE 141 — <character string literal>s are allowed to be zero-length strings (i.e., to contain no characters) even though it is not permitted to declare a <data type> that is CHARACTER with <character length> 0 (zero).
```

The note above implies that a character type with length 0 cannot be declared, but it doesn't say whether it can be derived or what the type of a zero-length string is supposed to be. As an aside, it says that the type of a literal string is CHAR(n), not VARCHAR(n).

Regarding concatenation, yes, it's effectively the sum of sizes of each of the types involved, capped at the maximum implementation-defined size for a character type.

**Comment by findepi** [comment]
@martint does it mean that `''` should be of type `char(0)` but the user should not be able to refer to such a type explicitly?
But then, CTAS with `''` would expose the `char(0)` type explicitly in the new table's schema.

**Comment by martint** [comment]
That's what's not entirely clear. I'm doing some more research using other sources.

**Comment by martint** [comment]
This confirms my suspicion. From the SQL 1999 book by Jim Melton:

> Sometimes, you may need to specify a character string literal with no characters in it (not even blank). SQL:1999 permits this; you just use two consecutive apostrophes (`''`) and the character string literal's data type is `CHARACTER`, but with length 0 (which is not a permitted data type for a column definition)
